Use SQL update rather than delete and insert for updating resources

deepankarb opened the pull request:
* Change from delete and reinsert to updating the rows via query
* Change non-existent resource update test to expect no insertion

jingtang10 commented on fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:
```diff
@@ -97,12 +98,16 @@ class DatabaseImplTest {
-        )
+        val resourceNotFoundInDbException = assertThrows(ResourceNotFoundInDbException::class.java,
+                { database.select(Patient::class.java, TEST_PATIENT_2_ID) })
+        Truth.assertThat(resourceNotFoundInDbException.message)
+                .isEqualTo("Resource not found with type " +
+                        TEST_PATIENT_2.resourceType.name +
+                        " and id " +
+                        TEST_PATIENT_2_ID +
+                        "!")
```
nit

```suggestion
        Truth.assertThat(resourceNotFoundInDbException.message).isEqualTo(
            "Resource not found with type ${TEST_PATIENT_2.resourceType.name} and id ${TEST_PATIENT_2_ID}!"
        )
```

jingtang10 commented on fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:
```diff
@@ -97,12 +98,16 @@ class DatabaseImplTest {
-            TEST_PATIENT_2,
-            database.select(Patient::class.java, TEST_PATIENT_2_ID)
-        )
+        val resourceNotFoundInDbException = assertThrows(ResourceNotFoundInDbException::class.java,
+                { database.select(Patient::class.java, TEST_PATIENT_2_ID) })
```
nit on indentation

```suggestion
        val resourceNotFoundInDbException = assertThrows(
            ResourceNotFoundInDbException::class.java,
            { database.select(Patient::class.java, TEST_PATIENT_2_ID) }
        )
```

jingtang10 commented on fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:
```diff
@@ -97,12 +98,16 @@ class DatabaseImplTest {
 
     @Test
-    fun update_nonExistingResource_shouldInsertResource() {
+    fun update_nonExistingResource_shouldNotInsertResource() {
```
```suggestion
    fun update_nonExistingResource_shouldThrowResourceNotFoundInDbException() {
```

deepankarb replied:
We are not throwing on `ResourceDao.update()` because the underlying raw query does not update any rows. To verify that no rows were modified we load the non-existent resource which throws. It's not the update mewthod which is throwing. Should we still name it as suggested?

jingtang10 commented on fhirengine/src/main/java/com/google/fhirengine/db/impl/dao/ResourceDao.kt:
```diff
@@ -46,8 +46,9 @@ internal abstract class ResourceDao {
     open fun update(resource: Resource) {
-        deleteResource(resource.id, resource.resourceType)
-        insert(resource)
+        updateResource(resource.id,
+                resource.resourceType,
+                iParser.encodeResourceToString(resource))
```
```suggestion
        updateResource(
            resource.id,
            resource.resourceType,
            iParser.encodeResourceToString(resource)
        )
```

jingtang10 commented on fhirengine/src/test-common/java/com/google/fhirengine/resource/TestingUtils.kt:
```diff
@@ -17,17 +17,17 @@
-        assertEquals(iParser.encodeResourceToString(expected),
-                iParser.encodeResourceToString(actual))
+        Truth.assertThat(iParser.encodeResourceToString(actual)).isEqualTo(
+                iParser.encodeResourceToString(expected))
```
nit: always leave the closing bracket on a new line

more examples https://kotlinlang.org/docs/reference/coding-conventions.html

deepankarb replied:
Moved entire method call to next line.

jingtang10 commented:
Thanks for the fast change @deepankarb!

I think you'll need to re-index the reource in the ResourceDao.kt

deepankarb commented:
> We found a Contributor License Agreement for you (the sender of this pull request), but were unable to find agreements for all the commit author(s) or Co-authors. If you authored these, maybe you used a different email address in the git commits than was used to sign the CLA ([login here](https://cla.developers.google.com/) to double check)? If these were authored by someone else, then they will need to sign a CLA as well, and confirm that they're okay with these being contributed to Google.
> In order to pass this check, please resolve this problem and then comment `@googlebot I fixed it.`. If the bot doesn't comment, it means it doesn't think anything has changed.
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Fgoogle%2Fandroid-fhir%2Fpull%2F160) for more info**.

@googlebot I fixed it.

jingtang10 commented on fhirengine/src/test-common/java/com/google/fhirengine/resource/TestingUtils.kt:
```diff
@@ -17,17 +17,17 @@
-        assertEquals(iParser.encodeResourceToString(expected),
-                iParser.encodeResourceToString(actual))
+        Truth.assertThat(iParser.encodeResourceToString(actual))
+                .isEqualTo(iParser.encodeResourceToString(expected))
```
nit: odd indentation. can you change to 4?

deepankarb replied:
I thought I did that. Will fix.

jingtang10 commented on fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:
```diff
@@ -97,11 +98,16 @@ class DatabaseImplTest {
+            { database.select(Patient::class.java, TEST_PATIENT_2_ID) }
+        )
+        Truth.assertThat(resourceNotFoundInDbException.message).isEqualTo(
+            /* ktlint-disable max-line-length */
+            "Resource not found with type ${TEST_PATIENT_2.resourceType.name} and id $TEST_PATIENT_2_ID!"
+            /* ktlint-enable max-line-length */
```
oh sorry feel free to put this in 2 lines. i didn't realise my suggestion was too long.

deepankarb replied:
I think this is better since the other option is harder to write/maintain even in an IDE.

jingtang10 commented on fhirengine/src/androidTest/java/com/google/fhirengine/db/impl/DatabaseImplTest.kt:
```diff
@@ -119,8 +125,8 @@ class DatabaseImplTest {
-    fun select_nonexistentResource_shouldThrowResourceNotFondException() {
-        val resourceNotFoundInDbException =
+    fun select_nonexistentResource_shouldThrowResourceNotFoundException() {
+        val resourceNotFoundException =
```
can this val be removed?

deepankarb replied:
I don't think so.

jingtang10 commented on fhirengine/src/test/java/com/google/fhirengine/impl/FhirEngineImplTest.kt:
```diff
@@ -76,12 +77,17 @@ class FhirEngineImplTest {
+        val resourceNotFoundException = assertThrows(ResourceNotFoundException::class.java,
+                { fhirEngine.load(Patient::class.java, TEST_PATIENT_2_ID) })
+        Truth.assertThat(resourceNotFoundException.message)
+                .isEqualTo("Resource not found with type " +
+                        TEST_PATIENT_2.resourceType.name +
+                        " and id " +
+                        TEST_PATIENT_2_ID +
+                        "!"
```
use the same kotlin string format (with $ signs) as the previous comment.
